File: week01/재귀함수/1914.py
# 이동 횟수가 최소인 경우
# K : 이동 횟수, A, B는 이동 과정
# N 이 20보다 크면 과정 출력 필요 없음

# hanoi(N, start, end, sub) => N개의 원판을 start에서 end로 이동
# 이 과정에서 1 -> 2 로 n - 1 만큼 이동
# n 번째 원판을 1 -> 3 으로 이동
# 다시 2에 있던 n - 1 의 원판을 2 -> 3 이동
N = int(input())
moving_count = 2 ** N - 1

def move(start, end):
  print(start, end)

# start : 1, end :3 sub : 2
def hanoi(N, start, end, sub):
  if N == 1:
    move(start, end)
  else:
    hanoi(N - 1, start, sub, end)
    move(start, end)
    hanoi(N - 1, sub, end, start)

if N <= 20:
  print(moving_count)
  hanoi(N, 1, 3, 2)
else:
  print(moving_count)

File: week01_JS/재귀함수/1914.js
// 하노이탑
// hanoi(3) => 2개의 원판 A -> B (C 경유)
// 3번째 원판 A -> C 
// 나머지 2개의 원판 B -> C (A 경유)

const fs = require('fs')
const input = fs.readFileSync('input.txt').toString().trim()

let N = parseInt(input)

// console.log는 시간 초과가 날 수 있다
let result = []; // 결과 출력용 배열

function hanoi(N, start, end, sub) {
  if (N == 1) {
    result.push(`${start} ${end}`)
  } else {
    hanoi(N - 1, start, sub, end)
    result.push(`${start} ${end}`)
    hanoi(N - 1, sub, end, start)
  }
}

console.log((BigInt(2) ** BigInt(N) - BigInt(1)).toString());

if (N <= 20) {
  hanoi(N, 1, 3, 2);
  console.log(result.join('\n'))
}